### PR TITLE
Messaging code for Streams

### PR DESCRIPTION
**Pull request by dtillman**

### CODE REVIEW

File: src/Messaging/Messaging.sln [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{D77F3F56-2B76-45CC-A3DF-6B4C46F09546}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "config", "config", "{7B896FC9-32DD-4A89-B09B-1E34674CEFEC}"
+	ProjectSection(SolutionItems) = preProject
+		..\..\nuget.config = ..\..\nuget.config
+		..\..\sharedproject.props = ..\..\sharedproject.props
+		..\..\sharedtest.props = ..\..\sharedtest.props
+		..\..\stylecop.json = ..\..\stylecop.json
+		..\..\targetframework.props = ..\..\targetframework.props
```
Comment (TimHess) [inline src/Messaging/Messaging.sln]:
```suggestion
```

File: src/Messaging/src/Abstractions/Core/IMessageRequestReplyOperations.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Steeltoe.Messaging.Core
+{
+    /// <summary>
+    /// Operations for sending messages to and receiving the reply from a destination.
+    /// </summary>
+    /// <typeparam name="D">type type of the destination</typeparam>
```
Comment (TimHess) [inline src/Messaging/src/Abstractions/Core/IMessageRequestReplyOperations.cs]:
```suggestion
    /// <typeparam name="D">type of the destination</typeparam>
```

File: src/Messaging/src/Abstractions/Handler/IMessageCondition.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        /// <param name="message">the message under process</param>
+        /// <returns>a condition instance in case of a match; or null if no match</returns>
+        T GetMatchingCondition(IMessage message);
+
+        /// <summary>
+        /// Compare this condition to another in the context of a specific message.
+        /// </summary>
+        /// <param name="other">the other condition to compare to</param>
+        /// <param name="message">the message under process</param>
+        /// <returns>results of the comparision</returns>
```
Comment (TimHess) [inline src/Messaging/src/Abstractions/Handler/IMessageCondition.cs]:
```suggestion
        /// <returns>results of the comparison</returns>
```

File: src/Messaging/Messaging.sln [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Messaging.Abstractions", "src\Abstractions\Steeltoe.Messaging.Abstractions.csproj", "{482680D5-D36C-4E63-9E43-D35B00492BA5}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Messaging.Base", "src\Base\Steeltoe.Messaging.Base.csproj", "{2C177BBD-9DAB-4926-BDDD-6A94C2F1DAE9}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Messaging.Base.Test", "test\Base.Test\Steeltoe.Messaging.Base.Test.csproj", "{80DC697C-BA47-4F25-A741-FD5464C71F06}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ChannelBenchmark", "test\Benchmarks\Channel\ChannelBenchmark.csproj", "{5F0DAE03-0CBD-4576-ACDD-0026A82E7760}"
+EndProject
+Global
```
Comment (TimHess) [inline src/Messaging/Messaging.sln]:
```suggestion
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "dependencies", "dependencies", "{2134CB62-4A32-4A3F-9AA5-C06B603D38FB}"
EndProject
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Common", "..\Common\src\Common\Steeltoe.Common.csproj", "{E5777518-72D9-4AB5-815C-AF9474BBFDEE}"
EndProject
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Common.TestResources", "..\Common\test\Common.TestResources\Steeltoe.Common.TestResources.csproj", "{9EA5A7BF-D6CF-4241-B32F-CDEAA93197B9}"
EndProject
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Steeltoe.Common.Abstractions", "..\Common\src\Abstractions\Steeltoe.Common.Abstractions.csproj", "{B0931691-6632-41C0-8175-539885BD5775}"
EndProject
Global
```

File: src/Messaging/Messaging.sln [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		{2C177BBD-9DAB-4926-BDDD-6A94C2F1DAE9}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{2C177BBD-9DAB-4926-BDDD-6A94C2F1DAE9}.Release|Any CPU.Build.0 = Release|Any CPU
+		{80DC697C-BA47-4F25-A741-FD5464C71F06}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{80DC697C-BA47-4F25-A741-FD5464C71F06}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{80DC697C-BA47-4F25-A741-FD5464C71F06}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{80DC697C-BA47-4F25-A741-FD5464C71F06}.Release|Any CPU.Build.0 = Release|Any CPU
+		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760}.Release|Any CPU.Build.0 = Release|Any CPU
```
Comment (TimHess) [inline src/Messaging/Messaging.sln]:
```suggestion
		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760}.Release|Any CPU.Build.0 = Release|Any CPU
		{E5777518-72D9-4AB5-815C-AF9474BBFDEE}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{E5777518-72D9-4AB5-815C-AF9474BBFDEE}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{E5777518-72D9-4AB5-815C-AF9474BBFDEE}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{E5777518-72D9-4AB5-815C-AF9474BBFDEE}.Release|Any CPU.Build.0 = Release|Any CPU
		{9EA5A7BF-D6CF-4241-B32F-CDEAA93197B9}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{9EA5A7BF-D6CF-4241-B32F-CDEAA93197B9}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{9EA5A7BF-D6CF-4241-B32F-CDEAA93197B9}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{9EA5A7BF-D6CF-4241-B32F-CDEAA93197B9}.Release|Any CPU.Build.0 = Release|Any CPU
		{B0931691-6632-41C0-8175-539885BD5775}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
		{B0931691-6632-41C0-8175-539885BD5775}.Debug|Any CPU.Build.0 = Debug|Any CPU
		{B0931691-6632-41C0-8175-539885BD5775}.Release|Any CPU.ActiveCfg = Release|Any CPU
		{B0931691-6632-41C0-8175-539885BD5775}.Release|Any CPU.Build.0 = Release|Any CPU

```

File: src/Messaging/Messaging.sln [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+	GlobalSection(NestedProjects) = preSolution
+		{482680D5-D36C-4E63-9E43-D35B00492BA5} = {6B21DB1B-B017-4432-8F04-55CC34DE524C}
+		{2C177BBD-9DAB-4926-BDDD-6A94C2F1DAE9} = {6B21DB1B-B017-4432-8F04-55CC34DE524C}
+		{80DC697C-BA47-4F25-A741-FD5464C71F06} = {D77F3F56-2B76-45CC-A3DF-6B4C46F09546}
+		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760} = {D77F3F56-2B76-45CC-A3DF-6B4C46F09546}
```
Comment (TimHess) [inline src/Messaging/Messaging.sln]:
```suggestion
		{5F0DAE03-0CBD-4576-ACDD-0026A82E7760} = {D77F3F56-2B76-45CC-A3DF-6B4C46F09546}
		{E5777518-72D9-4AB5-815C-AF9474BBFDEE} = {2134CB62-4A32-4A3F-9AA5-C06B603D38FB}
		{9EA5A7BF-D6CF-4241-B32F-CDEAA93197B9} = {2134CB62-4A32-4A3F-9AA5-C06B603D38FB}
		{B0931691-6632-41C0-8175-539885BD5775} = {2134CB62-4A32-4A3F-9AA5-C06B603D38FB}
```

File: src/Messaging/test/Base.Test/Steeltoe.Messaging.Base.Test.csproj [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    <PackageReference Include="Moq" Version="4.13.1" />
+    <PackageReference Include="xunit" Version="2.4.1" />
+    <PackageReference Include="xunit.runner.console" Version="2.4.1">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>-->
```
Comment (TimHess) [inline src/Messaging/test/Base.Test/Steeltoe.Messaging.Base.Test.csproj]: can be deleted

File: src/Messaging/test/Benchmarks/Channel/BenchmarkDotNet.Artifacts/results/Program-report-github.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+``` ini
```
Comment (TimHess) [inline src/Messaging/test/Benchmarks/Channel/BenchmarkDotNet.Artifacts/results/Program-report-github.md]: Should benchmark artifacts be included?

**Comment by TimHess** [comment]
`Steeltoe.Messaging.Base` doesn't follow the existing package naming convention, should it be `Steeltoe.MessagingBase`?

File: src/Messaging/test/Base.Test/Steeltoe.Messaging.Base.Test.csproj (hunk 7 above)
Comment (dtillman) [reply src/Messaging/test/Base.Test/Steeltoe.Messaging.Base.Test.csproj]: Yup

File: src/Messaging/test/Benchmarks/Channel/BenchmarkDotNet.Artifacts/results/Program-report-github.md (hunk 8 above)
Comment (dtillman) [reply src/Messaging/test/Benchmarks/Channel/BenchmarkDotNet.Artifacts/results/Program-report-github.md]: I think its worth having a record/trail of the benchmark performance for historical reasons.  I think it records the processor it was run on, etc... so might be helpful in the future...
But, I'm not hard set on it..

**Pull request merged**